### PR TITLE
Explicitly mark modules to be Safe

### PR DESCRIPTION
**Pull request by phadej**
Last PR for https://github.com/ekmett/lens/issues/959

### CODE REVIEW

File: src/Control/Monad/Free/TH.hs [hunk 1]
```diff
@@ -3,7 +3,11 @@
 {-# OPTIONS_GHC -Wno-overlapping-patterns #-}
 #endif
 #include "free-common.h"
-
+#if __GLASGOW_HASKELL__ >= 704 && MIN_VERSION_template_haskell(2,12,0)
```
Comment (RyanGlScott) [inline src/Control/Monad/Free/TH.hs:6]: `free` doesn't support any versions of GHC before 7.4, so the `__GLASGOW_HASKELL__ >= 704` part is redundant.

File: free.cabal [hunk 2]
```diff
@@ -81,25 +81,40 @@ library
...
+    indexed-traversable  >= 0.1.1   && < 0.2,
+    semigroupoids        >= 5.3.5   && < 6,
+    th-abstraction       >= 0.4.2.0 && < 0.5,
+    transformers         >= 0.3     && < 0.6,
+    transformers-base    >= 0.4.5.2 && < 0.5,
     template-haskell     >= 2.7.0.0 && < 2.18
 
+  -- GHC-7.8 bundles transformers-0.3,
+  -- mtl-2.2.* requires transformers >=0.4
+  if impl(ghc >=7.10)
```
Comment (RyanGlScott) [inline free.cabal:98]: What is the benefit of having a conditional dependency where both branches of the `if` declare a dependency on the same package? I would hope that `cabal`'s constraint solver would be smart enough to pick the appropriate version of `mtl` depending on which GHC is being used. But perhaps there's something deeper here that I'm missing.

File: free.cabal [hunk 3]
```diff
@@ -81,25 +81,40 @@ library
...
+    build-depends:
+      mtl               >= 2.2.2 && < 2.3
+  else
+    build-depends:
+      mtl               >= 2.1.3.1 && < 2.3
+
+  -- recent profunctors dropped support for GHCs older than 7.8
+  if impl(ghc >=7.8)
```
Comment (RyanGlScott) [inline free.cabal:106]: Similarly, this use of `if` confuses me.

File: src/Control/Alternative/Free.hs [hunk 4]
```diff
@@ -7,6 +7,12 @@
 #endif
 {-# OPTIONS_GHC -Wall #-}
 #include "free-common.h"
+-- Manual Typeable instances
+#if __GLASGOW_HASKELL__ < 707
```
Comment (RyanGlScott) [inline src/Control/Alternative/Free.hs:11]: Nitpick: I would prefer to invert the conditional here:

```hs
#if __GLASGOW_HASKELL__ >= 708
{-# LANGUAGE DeriveDataTypeable #-}
{-# LANGUAGE Safe #-}
#else
-- Manual Typeable instances
{-# LANGUAGE Trustworthy #-}
#endif
```

Notice that I'm folding the `DeriveDataTypeable` pragma above into this, as it belongs with the use of `Safe`.

Similar comments apply to other conditional uses of `Trustworthy`/`Safe` elsewhere.

File: free.cabal (hunk 2 above)
Comment (phadej) [reply free.cabal:98]: It's more restricted for older GHCs. Specifically, `mtl-2.2.1` isn't *explicitly safe*, so causes warnings. The dep definitions here are a bit complicated, but excludes (nasty, no-one should need) corner-cases.

File: free.cabal (hunk 3 above)
Comment (phadej) [reply free.cabal:106]: Here the same. For recent GHCs I want have a very tight (just released) lower bound.

FWIW, I think that GHC-7.8 support could be dropped above `profunctors`, it's in `lens`, so there is very little stuff in between. But that's something for later.

File: src/Control/Alternative/Free.hs (hunk 4 above)
Comment (phadej) [reply src/Control/Alternative/Free.hs:11]: Could I remove the `{-# OPTIONS_GHC -Wall #-}`, I think they don't add much value in source files.

Comment (RyanGlScott) [reply src/Control/Alternative/Free.hs:11]: I didn't even notice that!

I'd be fine with removing `-Wall` there (as I believe it's redundant), although perhaps that's better fit for a separate commit (in the same PR).

Comment (phadej) [reply src/Control/Alternative/Free.hs]: Bundled it all together. The diff is quite messy (preserving `OPTIONS_GHC -Wall` wouldn't make it any prettier).

File: src/Control/Monad/Free/TH.hs (hunk 1 above)
Comment (phadej) [reply src/Control/Monad/Free/TH.hs]: done.

File: src/Control/Monad/Free/Class.hs [hunk 5]
```diff
@@ -2,14 +2,15 @@
...
+{-# LANGUAGE Safe #-}
 {-# LANGUAGE UndecidableInstances #-}
 #if defined(__GLASGOW_HASKELL__) && __GLASGOW_HASKELL__ >= 704
 {-# LANGUAGE DefaultSignatures #-}
 {-# LANGUAGE TypeFamilies #-}
 #endif
 {-# OPTIONS_GHC -fno-warn-deprecations #-}
+{-# LANGUAGE Safe #-}
 #include "free-common.h"
-
```
Comment (RyanGlScott) [inline src/Control/Monad/Free/Class.hs:12]: Was this whitespace-only change intended?

File: src/Control/Monad/Free/TH.hs [hunk 6]
```diff
@@ -3,7 +3,11 @@
 {-# OPTIONS_GHC -Wno-overlapping-patterns #-}
 #endif
 #include "free-common.h"
```
Comment (RyanGlScott) [inline src/Control/Monad/Free/TH.hs:5]: Nitpick: I'd prefer `#include "free-common.h"` to come after all of the language pragmas, for consistency with the rest of the library.

File: src/Control/Comonad/Cofree.hs [hunk 7]
```diff
@@ -7,8 +7,13 @@
 {-# LANGUAGE DeriveDataTypeable #-}
 {-# LANGUAGE DeriveGeneric #-}
 {-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE Safe #-}
+#else
+-- Manual Typeable instances
+{-# LANGUAGE Trustworthy #-}
 #endif
 #include "free-common.h"
+
```
Comment (RyanGlScott) [inline src/Control/Comonad/Cofree.hs:16]: Was this whitespace-only change intended?

File: src/Control/Monad/Free/TH.hs (hunk 6 above)
Comment (phadej) [reply src/Control/Monad/Free/TH.hs]: Done

File: src/Control/Comonad/Cofree.hs (hunk 7 above)
Comment (phadej) [reply src/Control/Comonad/Cofree.hs:16]: Yes. I went through the modules, and now all of them have an empty line between language pragmas and the module header comment.

File: src/Control/Monad/Free/Class.hs (hunk 5 above)
Comment (phadej) [reply src/Control/Monad/Free/Class.hs:12]: This one wasn't

**Comment by phadej** [comment]
It would be great to have `free` released as well, so the table in https://github.com/ekmett/lens/issues/959 can be completed and the issue closed.

**Comment by RyanGlScott** [comment]
I've uploaded [`free-5.1.6`](http://hackage.haskell.org/package/free-5.1.6) to Hackage with these changes.

**Pull request merged**